[#5] Close AMQP Connections.

b-abel opened the pull request:
The MQTT endpoints are tracked for each tenant. When the last endpoint of a tenant is
closed, the AMQP connection for this tenant is closed as well.

b-abel committed:
Add test for failed message send.

Signed-off-by: Abel Buechner-Mihaljevic <abel.buechner-mihaljevic@bosch.io>

b-abel committed:
[#5] Close AMQP connection when all devices are disconnected.

The MQTT endpoints are tracked for each tenant. When the last endpoint of a tenant is
closed, the AMQP connection for this tenant is closed as well.

Signed-off-by: Abel Buechner-Mihaljevic <abel.buechner-mihaljevic@bosch.io>

calohmn commented on protocol-gateway/mqtt-protocol-gateway-template/src/main/java/org/eclipse/hono/gateway/sdk/mqtt2amqp/MultiTenantConnectionManager.java:
```diff
@@ -0,0 +1,118 @@
+    boolean closeEndpoint(String tenantId, MqttEndpoint mqttEndpoint);
+
+    /**
+     * Closes all connections, MQTT connections as well as AMQP for all tenants.
```
AMQP *connections*

calohmn commented on protocol-gateway/mqtt-protocol-gateway-template/src/test/java/org/eclipse/hono/gateway/sdk/mqtt2amqp/TenantConnectionsTest.java:
```diff
@@ -0,0 +1,118 @@
+package org.eclipse.hono.gateway.sdk.mqtt2amqp;
```
Missing copyright header.

b-abel replied:
Ouch! I needed to configure Checkstyle in my IDE explicitly to check test sources as well on this project :see_no_evil:

calohmn commented on protocol-gateway/mqtt-protocol-gateway-template/src/test/java/org/eclipse/hono/gateway/sdk/mqtt2amqp/MultiTenantConnectionManagerImplTest.java:
```diff
@@ -0,0 +1,121 @@
+package org.eclipse.hono.gateway.sdk.mqtt2amqp;
```
Missing copyright header.

calohmn commented on protocol-gateway/mqtt-protocol-gateway-template/src/test/java/org/eclipse/hono/gateway/sdk/mqtt2amqp/AbstractMqttProtocolGatewayTest.java:
```diff
@@ -579,6 +581,43 @@ public void testEventMessage() {
+
+        verify(protonSender).send(any(), handlerArgumentCaptor.capture());
+
+        ProtonDelivery protonDelivery = mock(ProtonDelivery.class);
```
final

calohmn commented on protocol-gateway/mqtt-protocol-gateway-template/src/test/java/org/eclipse/hono/gateway/sdk/mqtt2amqp/AbstractMqttProtocolGatewayTest.java:
```diff
@@ -494,27 +498,25 @@ public void testConfiguredCredentialsTakePrecedenceOverImplementation() {
-                // THEN the AMQP connection is authenticated with the configured credentials...
-                assertThat(clientConfig.getUsername()).isEqualTo(username);
-                assertThat(clientConfig.getPassword()).isEqualTo(password);
+        ArgumentCaptor<ClientConfigProperties> configPropertiesArgumentCaptor = ArgumentCaptor
```
final

calohmn commented on protocol-gateway/mqtt-protocol-gateway-template/src/test/java/org/eclipse/hono/gateway/sdk/mqtt2amqp/AbstractMqttProtocolGatewayTest.java:
```diff
@@ -455,26 +460,25 @@ public void testConnectWithGatewayCredentialsResolvedDynamicallySucceeds() {
-                // THEN the AMQP connection is authenticated with the provided credentials...
-                assertThat(clientConfig.getUsername()).isEqualTo(GW_USERNAME);
-                assertThat(clientConfig.getPassword()).isEqualTo(GW_PASSWORD);
+        ArgumentCaptor<ClientConfigProperties> configPropertiesArgumentCaptor = ArgumentCaptor
```
final

calohmn commented on protocol-gateway/mqtt-protocol-gateway-template/src/test/java/org/eclipse/hono/gateway/sdk/mqtt2amqp/AbstractMqttProtocolGatewayTest.java:
```diff
@@ -14,6 +14,8 @@
 
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.eclipse.hono.gateway.sdk.mqtt2amqp.TestMqttProtocolGateway.GW_PASSWORD;
+import static org.eclipse.hono.gateway.sdk.mqtt2amqp.TestMqttProtocolGateway.GW_USERNAME;
```
In Hono, such static member imports aren't used, I think we shouldn't use them here as well.

calohmn commented on protocol-gateway/mqtt-protocol-gateway-template/src/main/java/org/eclipse/hono/gateway/sdk/mqtt2amqp/TenantConnections.java:
```diff
@@ -0,0 +1,115 @@
+ * <p>
+ * Note: do not re-use an instance if it is already closed.
+ */
+class TenantConnections {
```
Missing javadoc on public methods here. Exception info on `getAmqpAdapterClientFactory()` javadoc will be helpful.

calohmn commented on protocol-gateway/mqtt-protocol-gateway-template/src/main/java/org/eclipse/hono/gateway/sdk/mqtt2amqp/MultiTenantConnectionManagerImpl.java:
```diff
@@ -0,0 +1,141 @@
+    private TenantConnections getTenantConnections(final String tenantId) throws IllegalArgumentException {
+        final TenantConnections tenantConnections = connectionsPerTenant.get(tenantId);
+        if (tenantConnections == null) {
+            throw new IllegalArgumentException("tenant [" + tenantId + "] is not connected");
```
Throwing an exception here doesn't look that appropriate to me because it requires exception handling in all calling methods and would also require mention in the `MultiTenantConnectionManager` interface javadoc. 
How about returning a `Future<TenantConnections>` here and letting it fail if the map entry is `null`.
The `MultiTenantConnectionManager#addEndpoint` and `MultiTenantConnectionManager#closeEndpoint` could then also return a Future, and the "not connected/no connection registered" case could be handled via a failed Future.

calohmn commented on protocol-gateway/mqtt-protocol-gateway-template/src/main/java/org/eclipse/hono/gateway/sdk/mqtt2amqp/MultiTenantConnectionManager.java:
```diff
@@ -0,0 +1,118 @@
+    Future<Void> connect(String tenantId, Vertx vertx, ClientConfigProperties clientConfig);
+
+    /**
+     * Adds an MQTT endpoint for the given tenant.
```
I think it would be good to mention that `connect` has to have been called here first for the tenant.

b-abel committed:
[#5] Improve JavaDoc.

Signed-off-by: Abel Buechner-Mihaljevic <abel.buechner-mihaljevic@bosch.io>

b-abel committed:
[#5] Conform CheckStyle rules (final keyword + license headers + static imports).

Signed-off-by: Abel Buechner-Mihaljevic <abel.buechner-mihaljevic@bosch.io>

b-abel committed:
[#5] Refactored MultiTenantConnectionManager.

MultiTenantConnectionManager.addEndpoint() and MultiTenantConnectionManager.closeEndpoint()
now return a future instead of throwing exceptions.

Closing TenantConnections no longer throws an exception if the instance is already closed.

Signed-off-by: Abel Buechner-Mihaljevic <abel.buechner-mihaljevic@bosch.io>

b-abel commented:
Added the requested changes.

calohmn commented on protocol-gateway/mqtt-protocol-gateway-template/src/main/java/org/eclipse/hono/gateway/sdk/mqtt2amqp/AbstractMqttProtocolGateway.java:
```diff
@@ -473,22 +467,29 @@ private void registerHandlers(final MqttEndpoint endpoint, final Device authenti
+        tenantConnectionManager.closeEndpoint(tenantId, endpoint)
+                .onSuccess(amqpLinkClosed -> {
+                    if (amqpLinkClosed) {
+                        log.info("closing AMQP connection for tenant [{}]", tenantId);
```
"closed" instead of "closing"?

calohmn commented on protocol-gateway/mqtt-protocol-gateway-template/src/main/java/org/eclipse/hono/gateway/sdk/mqtt2amqp/MultiTenantConnectionManager.java:
```diff
@@ -0,0 +1,125 @@
+public interface MultiTenantConnectionManager {
+
+    /**
+     * Connect to Hono's AMQP adapter with the given configuration.
```
"Connects [...]"

calohmn commented on protocol-gateway/mqtt-protocol-gateway-template/src/main/java/org/eclipse/hono/gateway/sdk/mqtt2amqp/MultiTenantConnectionManagerImpl.java:
```diff
@@ -0,0 +1,132 @@
+    }
+
+    private Future<AmqpAdapterClientFactory> getAmqpAdapterClientFactory(final String tenantId)
+            throws IllegalStateException, IllegalArgumentException {
```
It looks to me as if `IllegalArgumentException` can't be thrown here anymore.

calohmn commented on protocol-gateway/mqtt-protocol-gateway-template/src/main/java/org/eclipse/hono/gateway/sdk/mqtt2amqp/MultiTenantConnectionManagerImpl.java:
```diff
@@ -0,0 +1,132 @@
+    }
+
+    private Future<AmqpAdapterClientFactory> getAmqpAdapterClientFactory(final String tenantId)
+            throws IllegalStateException, IllegalArgumentException {
```
Concerning the `IllegalStateException` case:
Wouldn't it be better to return a failed Future here for that case? I guess, if this exception was thrown here, it would potentially not be handled properly up the call chain.

b-abel replied:
FMPOV the `IllegalStateException` indicates a programming error which should not be "handled", but considered a bug that needs to be fixed, analogous to the usage of `Objects.requireNonNull(obj)` in the Hono code.

calohmn replied:
Hmm, I don't know whether it's necessarily a programming error here. If there's a new MQTT connection being initiated for one device, causing `failedTenant.closeAllConnections();` to be invoked because of AMQP adapter connectivity isssue, and if the verticle event loop thread handles an MQTT Publish request from another device around the same time, it isn't immediately obvious to me, that the IllegalStateException can't happen.
Therefore I think, failing the Future and adding a WARN/ERROR log message instead would remove the need to be absolutely sure here, that such an edge case can't occur.

b-abel replied:
I followed your suggestions and additionally refactored some more methods in `TenantConnections` which now return futures as well to be more consistent.

b-abel committed:
[#5] Requested changes.

Some minor changes like log statements and JavaDoc.
+ Refactoring of some methods in TenantConnections to return futures to indicate success or failure.

Signed-off-by: Abel Buechner-Mihaljevic <abel.buechner-mihaljevic@bosch.io>

calohmn reviewed (approved):
LGTM
